OpenChange/MAPIStoreMailMessage.m

```diff
@@ -396,36 +401,82 @@ - (void) _fetchBodyData
             encoding = @"7-bit";
 
           /* We should provide a case for each of the types in acceptedMimeTypes */
-          if ([mimeType isEqualToString: @"text/html"] && !mailIsEvent)
-            {
-              content = [content bodyDataFromEncoding: encoding];
-              [htmlContent appendData: content];
-            }
-          else if ([mimeType isEqualToString: @"text/plain"] && !mailIsEvent)
+          if (!mailIsEvent)
             {
+              NSString *charset;
+              BOOL html;
+              if ([mimeType isEqualToString: @"text/html"])
+                {
+                  html = YES;
+                }
+              else if ([mimeType isEqualToString: @"text/plain"])
+                {
+                  html = NO;
+                }
+              else
+                {
+                  [self warnWithFormat: @"Unsupported MIME type for non-event body part: %@.",
+                        mimeType];
+                  continue;
+                }
+                          
               content = [content bodyDataFromEncoding: encoding];
-              NSString *charset = [bodyPartsCharsets objectForKey: key];
+              charset = [bodyPartsCharsets objectForKey: key];
               if (charset)
                 {
-                  NSString *stringValue = [content bodyStringFromCharset: charset];
-                  [textContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                  NSString *stringValue = nil;
+                  if (html)
+                    {
+                      if ([charset isEqualToString: headerCharset])
+                        {
+                          [htmlContent appendData: content];
+                        }
+                      else
+                        {
+                          stringValue = [content bodyStringFromCharset: charset];
+                          [htmlContent appendData: [stringValue dataUsingEncoding: NSUTF8StringEncoding]];
+                        }
+                      if (multipartMixed)
```

Comment:
There are two possibilities the charset is the same that the header charset of different.
If it is the same, no problem we add it.
If it is different then headerCharset is utf-8 so we convert to utf8 and add it. We know it is utf8 because we can only arrive to this situation if there are different charsets in the message and in this case we set it to utf8.